### PR TITLE
Straighten out types in FGG/RuleM/Compile

### PR DESCRIPTION
**Pull request by davidweichiang**

### CODE REVIEW

**Comment by davidweichiang** [comment]
This is a big set of changes, but the summary is:

- Instead of using Strings everywhere, define new types for Value, NodeName, NodeLabel, and EdgeLabel. These are no longer Strings, but structured objects.
- Remove the type parameter `d` from FGG, HGF, etc. NodeLabel (which is a synonym for Type) takes its place.
- A special case for creating a start rule was removed; I don't think it was necessary.
- The dangling edges created by `discardEdges'` were removed; I don't think they were necessary.

Hopefully it should be clearer now where `Var`s used (for #107).

File: src/Util/FGG.hs [hunk 1]
```diff
@@ -3,6 +3,7 @@
 module Util.FGG where
 import qualified Data.Map as Map
 import Data.List
+import Struct.Lib
```
Comment (davidweichiang) [inline src/Util/FGG.hs:6]: It's a little unfortunate to introduce this dependency

**Comment by ccshan** [comment]
I never really understood the `FGG` type so I expect to review whether I can understand this `FGG` type rather than the changes~

File: src/Util/FGG.hs [hunk 2]
```diff
@@ -12,73 +13,44 @@ import Util.JSON
...
+  show (ElNonterminal tm) = show tm
+  show (ElTerminal s) = s
+
+data Edge = Edge { edge_atts :: [(NodeName, NodeLabel)], edge_label :: EdgeLabel }
   deriving Eq
-data HGF d = HGF { hgf_nodes :: [Node d], hgf_edges :: [Edge], hgf_exts :: [Int]}
+data HGF = HGF { hgf_nodes :: [(NodeName, NodeLabel)], hgf_edges :: [Edge], hgf_exts :: [(NodeName, NodeLabel)] }
   deriving Eq
-data Edge' d = Edge' { edge_atts' :: [(Label, d)], edge_label' :: Label }
+data Rule = Rule EdgeLabel HGF
```
Comment (davidweichiang) [inline src/Util/FGG.hs:44]: Rule and HGF could be merged into a single type.

**Pull request merged**